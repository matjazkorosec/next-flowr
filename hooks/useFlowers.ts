import { useState, useEffect } from 'react';
import apiClient from '@/lib/apiClient';
import { Flower } from '@/types/flower';

const useFlowers = () => {
  const [flowers, setFlowers] = useState<Flower[]>([]);
  const [loading, setLoading] = useState<boolean>(true);

  useEffect(() => {
    const fetchFlowers = async () => {
      setLoading(true);
      try {
        const response = await apiClient.get<{ items: Flower[] }>('/flowers');
        setFlowers(response.data.items);
      } catch (err) {
        console.error('Error fetching flowers:', err);
      } finally {
        setLoading(false);
      }
    };

    fetchFlowers();
  }, []);

  return { flowers, loading };
};

export default useFlowers;
