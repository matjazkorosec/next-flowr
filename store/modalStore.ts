import { ModalState } from '@/types/modal';
import { create } from 'zustand';

export const useModalStore = create<ModalState>((set) => ({
  modalContent: null,
  showModal: (content) => set({ modalContent: content }),
  hideModal: () => set({ modalContent: null }),
}));
