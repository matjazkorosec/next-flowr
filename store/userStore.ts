import apiClient from '@/lib/apiClient';
import { UserState } from '@/types/user';
import { create } from 'zustand';
import { persist } from 'zustand/middleware';

export const useUserStore = create(
  persist<UserState>(
    (set, get) => ({
      isLoggedIn: false,
      user: null,
      accessToken: null,
      refreshToken: null,

      setTokens: ({ accessToken, refreshToken }) =>
        set(() => ({
          isLoggedIn: true,
          accessToken,
          refreshToken,
        })),

      getUserInfo: async () => {
        const { accessToken } = get();

        if (!accessToken) {
          console.error('No access token found');
          return;
        }

        try {
          const response = await apiClient.get('/account/me', {
            headers: {
              Authorization: `Bearer ${accessToken}`,
            },
          });

          const userData = response.data;

          set(() => ({
            user: {
              id: userData.id,
              email: userData.email,
              firstName: userData.firstName,
              lastName: userData.lastName,
              dateOfBirth: userData.dateOfBirth,
              sightingsNum: userData.sightingsNum,
              pictureUrl: userData.pictureUrl,
            },
          }));
        } catch (error) {
          console.error('Failed to fetch user info', error);
        }
      },

      logout: () => {
        set(() => ({
          isLoggedIn: false,
          user: null,
          accessToken: null,
          refreshToken: null,
        }));
      },
    }),
    {
      name: 'FlowrSpotUserStore',
      merge: (persistedState, currentState) => ({
        ...currentState,
        ...(persistedState || {}),
      }),
    }
  )
);
