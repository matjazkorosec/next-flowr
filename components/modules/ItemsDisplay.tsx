'use client';

import TempPhoto from '@/assets/flower.jpg';
import Image from 'next/image';
import StarIcon from '@/assets/icon/star.svg';
import StarIconActive from '@/assets/icon/star-active.svg';
import { useState } from 'react';
import useFlowers from '@/hooks/useFlowers';
import { Flower } from '@/types/flower';
import { Skeleton } from '@/components/ui/skeleton';
import { useUserStore } from '@/store/userStore';

const ItemsDisplay = () => {
  const [isFavorite, setIsFavorite] = useState<Record<string, boolean>>({});
  const { flowers, loading } = useFlowers();
  const { user } = useUserStore();

  const handleStarClick = (id: string) => {
    setIsFavorite((prev) => ({ ...prev, [id]: !prev[id] }));
  };

  if (loading) {
    return (
      <section className="items-display">
        <div className="adapt">
          <ul className="item-list">
            {Array.from({ length: 4 }).map((_, index) => (
              <li key={`skeleton-${index}`}>
                <Skeleton role="status" className="w-full h-[337px]" />
              </li>
            ))}
          </ul>
        </div>
      </section>
    );
  }

  return (
    <section className="items-display">
      <div className="adapt">
        <ul className="item-list">
          {flowers.map((flower: Flower) => (
            <li key={flower.id}>
              <figure>
                <Image
                  src={flower.pictureUrl || TempPhoto.src}
                  alt={flower.name}
                  width={270}
                  height={337}
                  style={{ width: '100%', height: 'auto' }}
                  priority
                />
                <div className="figcaption">
                  <h3>{flower.name}</h3>
                  <p>{flower.latinName}</p>
                  <span className="sightings-count">
                    {flower.sightingsNum} sightings
                  </span>
                </div>
              </figure>
              {user && (
                <button
                  className={`star ${isFavorite[flower.id] ? 'favorite' : ''}`}
                  onClick={() => handleStarClick(flower.id)}
                >
                  <span
                    style={{
                      backgroundImage: `url(${
                        isFavorite[flower.id]
                          ? StarIconActive.src
                          : StarIcon.src
                      })`,
                    }}
                  >
                    <span className="sr-only">
                      {isFavorite[flower.id]
                        ? 'Remove favorite'
                        : 'Make favorite'}
                    </span>
                  </span>
                </button>
              )}
            </li>
          ))}
        </ul>
      </div>
    </section>
  );
};

export default ItemsDisplay;
