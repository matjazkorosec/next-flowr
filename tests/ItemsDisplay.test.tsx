import { render, screen } from '@testing-library/react';
import ItemsDisplay from '@/components/modules/ItemsDisplay';
import useFlowers from '@/hooks/useFlowers';
import { useUserStore } from '@/store/userStore';

jest.mock('@/hooks/useFlowers', () => jest.fn());
jest.mock('@/store/userStore', () => ({
  useUserStore: jest.fn(),
}));

const mockFlowers = [
  { id: '1', name: 'Roza', sightingsNum: 2 },
  { id: '2', name: 'Kopriva', sightingsNum: 4 },
];

const mockUser = {
  id: '1',
  firstName: 'Mat',
  lastName: 'Kor',
  email: '',
  dateOfBirth: '',
  sightingsNum: 0,
  pictureUrl: '',
};

describe('ItemsDisplay Component', () => {
  beforeEach(() => {
    (useUserStore as unknown as jest.Mock).mockReturnValue({
      user: mockUser,
      logout: jest.fn(),
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Should display flowers when data is loaded', () => {
    (useFlowers as jest.Mock).mockReturnValue({
      flowers: mockFlowers,
      loading: false,
    });

    render(<ItemsDisplay />);

    expect(screen.getByText('Roza')).toBeInTheDocument();
    expect(screen.getByText('Kopriva')).toBeInTheDocument();
    expect(screen.getByText('2 sightings')).toBeInTheDocument();
    expect(screen.getByText('4 sightings')).toBeInTheDocument();
  });

  it('Should display skeletons during loading', () => {
    (useFlowers as jest.Mock).mockReturnValue({
      flowers: [],
      loading: true,
    });

    render(<ItemsDisplay />);

    const skeletons = screen.getAllByRole('status');
    expect(skeletons).toHaveLength(4);
  });

  it('Should display favorite button when user is logged in', () => {
    (useFlowers as jest.Mock).mockReturnValue({
      flowers: mockFlowers,
      loading: false,
    });

    render(<ItemsDisplay />);

    const favoriteButtons = screen.getAllByRole('button', {
      name: /make favorite/i,
    });
    expect(favoriteButtons).toHaveLength(2);
  });

  it('Should hide favorite button when user is logged out', () => {
    (useFlowers as jest.Mock).mockReturnValue({
      flowers: mockFlowers,
      loading: false,
    });

    (useUserStore as unknown as jest.Mock).mockReturnValue({
      user: null,
      logout: jest.fn(),
    });

    render(<ItemsDisplay />);

    const favoriteButtons = screen.queryAllByRole('button', {
      name: /make favorite/i,
    });
    expect(favoriteButtons).toHaveLength(0);
  });

  it('Should match the snapshot for loaded flowers', () => {
    (useFlowers as jest.Mock).mockReturnValue({
      flowers: mockFlowers,
      loading: false,
    });

    const { container } = render(<ItemsDisplay />);
    expect(container).toMatchSnapshot();
  });

  it('Should match the snapshot for no user logged in', () => {
    (useFlowers as jest.Mock).mockReturnValue({
      flowers: mockFlowers,
      loading: false,
    });

    (useUserStore as unknown as jest.Mock).mockReturnValue({
      user: null,
      logout: jest.fn(),
    });

    const { container } = render(<ItemsDisplay />);
    expect(container).toMatchSnapshot();
  });
});
