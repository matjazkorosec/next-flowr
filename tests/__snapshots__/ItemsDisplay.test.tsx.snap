// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemsDisplay Component Should match the snapshot for loaded flowers 1`] = `
<div>
  <section
    class="items-display"
  >
    <div
      class="adapt"
    >
      <ul
        class="item-list"
      >
        <li>
          <figure>
            <img
              alt="Roza"
              data-nimg="1"
              decoding="async"
              height="337"
              src="/_next/image?url=%2Fimg.jpg&w=640&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=384&q=75 1x, /_next/image?url=%2Fimg.jpg&w=640&q=75 2x"
              style="color: transparent; width: 100%; height: auto;"
              width="270"
            />
            <div
              class="figcaption"
            >
              <h3>
                Roza
              </h3>
              <p />
              <span
                class="sightings-count"
              >
                2
                 sightings
              </span>
            </div>
          </figure>
          <button
            class="star "
          >
            <span
              style="background-image: url(/img.jpg);"
            >
              <span
                class="sr-only"
              >
                Make favorite
              </span>
            </span>
          </button>
        </li>
        <li>
          <figure>
            <img
              alt="Kopriva"
              data-nimg="1"
              decoding="async"
              height="337"
              src="/_next/image?url=%2Fimg.jpg&w=640&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=384&q=75 1x, /_next/image?url=%2Fimg.jpg&w=640&q=75 2x"
              style="color: transparent; width: 100%; height: auto;"
              width="270"
            />
            <div
              class="figcaption"
            >
              <h3>
                Kopriva
              </h3>
              <p />
              <span
                class="sightings-count"
              >
                4
                 sightings
              </span>
            </div>
          </figure>
          <button
            class="star "
          >
            <span
              style="background-image: url(/img.jpg);"
            >
              <span
                class="sr-only"
              >
                Make favorite
              </span>
            </span>
          </button>
        </li>
      </ul>
    </div>
  </section>
</div>
`;

exports[`ItemsDisplay Component Should match the snapshot for no user logged in 1`] = `
<div>
  <section
    class="items-display"
  >
    <div
      class="adapt"
    >
      <ul
        class="item-list"
      >
        <li>
          <figure>
            <img
              alt="Roza"
              data-nimg="1"
              decoding="async"
              height="337"
              src="/_next/image?url=%2Fimg.jpg&w=640&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=384&q=75 1x, /_next/image?url=%2Fimg.jpg&w=640&q=75 2x"
              style="color: transparent; width: 100%; height: auto;"
              width="270"
            />
            <div
              class="figcaption"
            >
              <h3>
                Roza
              </h3>
              <p />
              <span
                class="sightings-count"
              >
                2
                 sightings
              </span>
            </div>
          </figure>
        </li>
        <li>
          <figure>
            <img
              alt="Kopriva"
              data-nimg="1"
              decoding="async"
              height="337"
              src="/_next/image?url=%2Fimg.jpg&w=640&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=384&q=75 1x, /_next/image?url=%2Fimg.jpg&w=640&q=75 2x"
              style="color: transparent; width: 100%; height: auto;"
              width="270"
            />
            <div
              class="figcaption"
            >
              <h3>
                Kopriva
              </h3>
              <p />
              <span
                class="sightings-count"
              >
                4
                 sightings
              </span>
            </div>
          </figure>
        </li>
      </ul>
    </div>
  </section>
</div>
`;
